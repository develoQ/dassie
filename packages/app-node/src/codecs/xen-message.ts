import { Reader, Writer } from "oer-utils"
import type { MarkOptional } from "ts-essentials"

const HELLO_NEIGHBOR_PROOF_LENGTH = 32

export type XenMessage = XenHelloMessage

export enum XenMessageType {
  Hello = 0,
}

export interface XenHelloMessage {
  method: XenMessageType.Hello
  signed: XenHelloMessageSignedPortion
  signature: Buffer
}

export interface XenHelloMessageSignedPortion {
  nodeId: string
  sequence: number
  neighbors: XenHelloNeighbor[]
}

export interface XenHelloNeighbor {
  nodeId: string
  proof: Buffer
}

export const parseMessage = (message: Buffer): XenMessage => {
  const reader = Reader.from(message)
  const method = reader.readUInt8Number()

  switch (method) {
    case XenMessageType.Hello: {
      const signed = reader.readVarOctetString()
      const signature = reader.readOctetString(64)
      const signedReader = Reader.from(signed)
      const nodeId = signedReader.readVarOctetString().toString("ascii")
      const sequence = signedReader.readUInt32Number()
      const neighborCount = signedReader.readUInt16Number()
      const neighbors: XenHelloNeighbor[] = Array.from({
        length: neighborCount,
      })

      for (let index = 0; index < neighborCount; index++) {
        const neighborDefinition = signedReader.readVarOctetString()
        const neighborReader = Reader.from(neighborDefinition)
        const nodeId = neighborReader.readVarOctetString().toString("ascii")
        const proof = neighborReader.readOctetString(
          HELLO_NEIGHBOR_PROOF_LENGTH
        )

        neighbors[index] = {
          nodeId,
          proof,
        }
      }

      return {
        method,
        signed: {
          nodeId,
          sequence,
          neighbors,
        },
        signature,
      }
    }
    default:
      throw new Error("Unknown Xen message type")
  }
}

export type XenMessageWithOptionalSignature = MarkOptional<
  XenMessage,
  "signature"
>

const writeSignature = (
  writer: Writer,
  message: XenMessageWithOptionalSignature,
  signed: Buffer,
  sign?: (content: Buffer) => Buffer
) => {
  if (message.signature != null) {
    writer.writeOctetString(message.signature, 64)
  } else if (sign != null) {
    writer.writeOctetString(sign(signed), 64)
  } else {
    throw new TypeError(
      "encodeMessage requires that a signing function is provided as the second parameter if message.signature is null"
    )
  }
}

export function encodeMessage(message: XenMessage): Buffer
export function encodeMessage(
  message: XenMessageWithOptionalSignature,
  sign?: (content: Buffer) => Buffer
): Buffer
export function encodeMessage(
  message: XenMessageWithOptionalSignature,
  sign?: (content: Buffer) => Buffer
): Buffer {
  const writer = new Writer()
  writer.writeUInt8(message.method)

  switch (message.method) {
    case XenMessageType.Hello: {
      const signedWriter = new Writer()

      signedWriter.writeVarOctetString(
        Buffer.from(message.signed.nodeId, "ascii")
      )
      signedWriter.writeUInt32(message.signed.sequence)
      signedWriter.writeUInt16(message.signed.neighbors.length)

      for (const neighbor of message.signed.neighbors) {
        const neighborWriter = new Writer()
        neighborWriter.writeVarOctetString(
          Buffer.from(neighbor.nodeId, "ascii")
        )
        neighborWriter.writeOctetString(
          neighbor.proof,
          HELLO_NEIGHBOR_PROOF_LENGTH
        )
        signedWriter.writeVarOctetString(neighborWriter.getBuffer())
      }

      const signed = signedWriter.getBuffer()

      writer.writeVarOctetString(signed)
      writeSignature(writer, message, signed, sign)

      return writer.getBuffer()
    }
  }
}
