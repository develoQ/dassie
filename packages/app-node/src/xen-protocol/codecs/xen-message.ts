import { Reader, Writer } from "oer-utils"
import type { MarkOptional } from "ts-essentials"

import {
  ia5String,
  octetString,
  sequence,
  sequenceOf,
  uint64Bigint,
} from "@xen-ilp/lib-oer"

const HELLO_NEIGHBOR_PROOF_LENGTH = 32

export type XenMessage = XenHelloMessage

export enum XenMessageType {
  Hello = 0,
}

export interface XenHelloMessage {
  method: XenMessageType.Hello
  signed: XenHelloMessageSignedPortion
  signature: Buffer
}

const xenHelloMessageSignedPortion = sequence({
  nodeId: ia5String(),
  sequence: uint64Bigint(),
  url: ia5String(),
  neighbors: sequenceOf(
    sequence({
      nodeId: ia5String(),
      proof: octetString({ fixedLength: HELLO_NEIGHBOR_PROOF_LENGTH }),
    })
  ),
})
export interface XenHelloMessageSignedPortion {
  nodeId: string
  sequence: bigint
  url: string
  neighbors: XenHelloNeighbor[]
}

export interface XenHelloNeighbor {
  nodeId: string
  proof: Uint8Array
}

export const parseMessage = (message: Uint8Array): XenMessage => {
  const reader = Reader.from(Buffer.from(message))
  const method = reader.readUInt8Number()

  switch (method) {
    case XenMessageType.Hello: {
      const signed = reader.readVarOctetString()
      const signature = reader.readOctetString(64)

      const signedPortionParseResult =
        xenHelloMessageSignedPortion.parse(signed)

      if (!signedPortionParseResult.success) {
        throw new Error("Failed to parse signed portion")
      }

      const { nodeId, sequence, url, neighbors } =
        signedPortionParseResult.value

      return {
        method,
        signed: {
          nodeId,
          sequence,
          url,
          neighbors,
        },
        signature,
      }
    }
    default:
      throw new Error("Unknown Xen message type")
  }
}

export type XenMessageWithOptionalSignature = MarkOptional<
  XenMessage,
  "signature"
>

const writeSignature = (
  writer: Writer,
  message: XenMessageWithOptionalSignature,
  signed: Buffer,
  sign?: (content: Buffer) => Buffer
) => {
  if (message.signature != null) {
    writer.writeOctetString(message.signature, 64)
  } else if (sign != null) {
    writer.writeOctetString(sign(signed), 64)
  } else {
    throw new TypeError(
      "encodeMessage requires that a signing function is provided as the second parameter if message.signature is null"
    )
  }
}

export function encodeMessage(message: XenMessage): Buffer
export function encodeMessage(
  message: XenMessageWithOptionalSignature,
  sign: (content: Buffer) => Buffer
): Buffer
export function encodeMessage(
  message: XenMessageWithOptionalSignature,
  sign?: (content: Buffer) => Buffer
): Buffer {
  const writer = new Writer()
  writer.writeUInt8(message.method)

  switch (message.method) {
    case XenMessageType.Hello: {
      const signed = xenHelloMessageSignedPortion.serialize(message.signed)

      if (!signed.success) {
        throw new Error("Failed to serialize signed portion")
      }

      writer.writeVarOctetString(Buffer.from(signed.value))
      writeSignature(writer, message, Buffer.from(signed.value), sign)

      return writer.getBuffer()
    }
  }
}
