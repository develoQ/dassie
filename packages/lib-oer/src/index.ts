export type { Infer, OerType, AnyOerType, ParseOptions } from "./base-type"
export * from "./bitstring"
export * from "./boolean"
export * from "./choice"
export * from "./enumerated"
export * from "./integer-as-bigint"
export * from "./integer-as-number"
export * from "./octet-string"
export * from "./sequence-of"
export * from "./sequence"
export * from "./string"
