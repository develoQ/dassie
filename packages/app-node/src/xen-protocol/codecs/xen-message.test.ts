import { describe, test } from "vitest"

import { hexToUint8Array } from "@xen-ilp/lib-oer"

import { parseMessage } from "./xen-message"

describe("xen message codec", () => {
  test("should parse a hello message", ({ expect }) => {
    const uint8Array = hexToUint8Array(
      `00 6d 02 6e 33 00 00 01 82 38 38 06 72 19 68 74 74 70 73 3a 2f 2f 6e 33 
       2e 6c 6f 63 61 6c 68 6f 73 74 3a 34 30 30 32 01 02 02 6e 31 00 00 00 00 
       00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
       00 00 00 00 02 6e 32 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
       00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 eb dc 55 15 35 8b 46 1a 1e 
       3a 40 3c f2 d7 25 21 1c ec 0a d2 c6 8a e3 e8 00 f1 85 bd 08 18 76 e7 be 
       db 43 e9 da 2f cf ed 18 52 63 d4 6d f0 53 62 bb 4a a1 df eb 87 97 73 93 
       8b bc 89 56 c0 55 0d`
    )

    const message = parseMessage(uint8Array)

    expect(message).toMatchSnapshot()
  })
})
