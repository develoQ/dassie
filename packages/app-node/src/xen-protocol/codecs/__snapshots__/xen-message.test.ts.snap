// Vitest Snapshot v1

exports[`xen message codec > should parse a hello message 1`] = `
{
  "method": 0,
  "signature": {
    "data": [
      235,
      220,
      85,
      21,
      53,
      139,
      70,
      26,
      30,
      58,
      64,
      60,
      242,
      215,
      37,
      33,
      28,
      236,
      10,
      210,
      198,
      138,
      227,
      232,
      0,
      241,
      133,
      189,
      8,
      24,
      118,
      231,
      190,
      219,
      67,
      233,
      218,
      47,
      207,
      237,
      24,
      82,
      99,
      212,
      109,
      240,
      83,
      98,
      187,
      74,
      161,
      223,
      235,
      135,
      151,
      115,
      147,
      139,
      188,
      137,
      86,
      192,
      85,
      13,
    ],
    "type": "Buffer",
  },
  "signed": {
    "neighbors": [
      {
        "nodeId": "n1",
        "proof": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          "type": "Buffer",
        },
      },
      {
        "nodeId": "n2",
        "proof": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          "type": "Buffer",
        },
      },
    ],
    "nodeId": "n3",
    "sequence": 1658800572018n,
    "url": "https://n3.localhost:4002",
  },
}
`;
